common/src/main/java/io/netty/util/DefaultAttributeMap.java

```diff
@@ -17,186 +17,240 @@
+                return;
             }
-            return false;
         }
     }
 
-    private static int index(AttributeKey<?> key) {
-        return key.id() & MASK;
-    }
-
     @SuppressWarnings("serial")
-    private static final class DefaultAttribute<T> extends AtomicReference<T> implements Attribute<T> {
+    private static final class DefaultAttribute<T> implements Attribute<T> {
+
+        private static final AtomicReferenceFieldUpdater<DefaultAttribute, Object> VALUE_UPDATER =
+                AtomicReferenceFieldUpdater.newUpdater(DefaultAttribute.class, Object.class, "value");
 
+        private static final Object REMOVED = new Object();
         private static final long serialVersionUID = -2661411462200283011L;
 
-        // The head of the linked-list this attribute belongs to
-        private final DefaultAttribute<?> head;
+        private final DefaultAttributeMap attributeMap;
         private final AttributeKey<T> key;
+        private volatile Object value;
 
-        // Double-linked list to prev and next node to allow fast removal
-        private DefaultAttribute<?> prev;
-        private DefaultAttribute<?> next;
+        DefaultAttribute(DefaultAttributeMap attributeMap,
+                         AttributeKey<T> key) {
+            this.attributeMap = attributeMap;
+            this.key = key;
+        }
 
-        // Will be set to true one the attribute is removed via getAndRemove() or remove()
-        private volatile boolean removed;
+        @Override
+        public AttributeKey<T> key() {
+            return key;
+        }
 
-        DefaultAttribute(DefaultAttribute<?> head, AttributeKey<T> key) {
-            this.head = head;
-            this.key = key;
+        @Override
+        public T get() {
+            Object v = value;
+            return v == REMOVED? null : (T) v;
         }
 
-        // Special constructor for the head of the linked-list.
-        DefaultAttribute() {
-            head = this;
-            key = null;
+        public boolean removed() {
+            return value == REMOVED;
         }
 
         @Override
-        public AttributeKey<T> key() {
-            return key;
+        public void set(T value) {
+            Object old;
+            do {
+                old = this.value;
+                if (old == REMOVED) {
+                    // a subsequent get should return null regardless the current value
+                    // given that remove has been called
+                    return;
+                }
+            } while (!VALUE_UPDATER.compareAndSet(this, old, value));
+        }
+
+        @Override
+        public T getAndSet(T value) {
+            Object old;
+            do {
+                old = this.value;
+                if (old == REMOVED) {
+                    // a subsequent get should return null regardless the current value
+                    // given that remove has been called
+                    return null;
```

Comment:
Is this the same as the old behavior? i.e. if the attribute was already removed from the map would its value still be set

Reply:
Nope but there was no documentation about it so I have changed it into something that would reduce the memory footprint of the instance while keeping a consistent behaviour

Reply:
Nope but there was no documentation about it so I have changed it into something that would reduce the memory footprint of the instance while keeping a consistent behaviour